### PR TITLE
model category exercise

### PR DESCRIPTION
**Pull request by singhal-ekansh**

### CODE REVIEW

File: app/models/category.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class Category < ApplicationRecord
+  has_many :sub_categories, class_name: "Category", foreign_key: "parent_id", dependent: :destroy
+  belongs_to :parent, class_name: "Category", optional: true
+
+  has_many :products, dependent: :restrict_with_error
+  has_many :sub_category_products,  through: :sub_categories,  source: :products
+
+  validates :name, presence: true
+  
+  validates_with UniqueRootCategoryName, if: :isNamePresent?
```
Comment (harsh-ag) [inline app/models/category.rb:10]: no need of custom validation here. use uniqueness with conditions

File: app/models/category.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  validates_with UniqueSubCategoryName, if: :isNamePresent?
```
Comment (harsh-ag) [inline app/models/category.rb:11]: same

File: app/models/category.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  validates_with CheckNestingLevel
```
Comment (harsh-ag) [inline app/models/category.rb:12]: rename

File: app/validators/check_nesting_level.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class CheckNestingLevel < ActiveModel::Validator
+  def validate(record)
+    return unless record.parent_id
+
+    sub_categories_ids = Category.where("parent_id is not null").pluck(:id)
```
Comment (harsh-ag) [inline app/validators/check_nesting_level.rb:5]: don't write query here. use associations. they are more readable.

File: app/models/category.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+class Category < ApplicationRecord
+  has_many :sub_categories, class_name: "Category", foreign_key: "parent_id", dependent: :destroy
+  belongs_to :parent, class_name: "Category", optional: true
+
+  has_many :products, dependent: :restrict_with_error
+  has_many :sub_category_products,  through: :sub_categories,  source: :products
```
Comment (harsh-ag) [inline app/models/category.rb:6]: dependent?

File: app/models/product.rb [hunk 6]
```diff
@@ -12,6 +13,26 @@ class Product < ApplicationRecord
     self.discount_price = price if discount_price.blank?
   end
 
+  after_create_commit do
```
Comment (harsh-ag) [inline app/models/product.rb:16]: do not pass block in the callbacks. create a method for it. DO NOT USE this practice to pass a block in callbacks

File: app/models/product.rb [hunk 7]
```diff
@@ -12,6 +13,26 @@ class Product < ApplicationRecord
...
+    root_category = Category.find(category_id).parent
+    root_category.increment!(:products_count) if root_category
```
Comment (harsh-ag) [inline app/models/product.rb:18]: will it always work? let's discuss.

File: app/controllers/store_controller.rb [hunk 8]
```diff
@@ -10,5 +10,9 @@ def index
       @products = Product.order(:title)
     end
   end
+
+  def categories
+    @categories = Category.includes(:sub_categories).where(parent_id: nil).all
```
Comment (harsh-ag) [inline app/controllers/store_controller.rb:15]: what will the difference b/w and which one is better?

`Category.includes(:parent)` vs `Category.includes(:sub_categories)`

**Review by harsh-ag — CHANGES_REQUESTED** [review]

File: app/models/category.rb (hunk 1 above)
Comment (singhal-ekansh) [reply app/models/category.rb:10]: validates :name, uniqueness: { scope: :parent_id, message: 'name of root category should be unique' }, allow_blank: true, unless: :isParentPresent?

File: app/models/category.rb (hunk 2 above)
Comment (singhal-ekansh) [reply app/models/category.rb:11]: validates :name, uniqueness: { scope: :parent_id, message: 'name of sub categories should be unique for each category' }, allow_blank: true, if: :isParentPresent?

File: app/models/category.rb (hunk 3 above)
Comment (singhal-ekansh) [reply app/models/category.rb:12]: NoChildOfSubCategory

**Pull request closed without merging**